vowpalwabbit/array_parameters.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef _WIN32
+#include <sys/mman.h>
+#endif
+
+using namespace std;
+typedef float weight;
+
+class weight_vector;
+
+class weights_iterator_iterator
+{
+private:
+	weight* _cur;
+public:
+	weights_iterator_iterator(weight* cur)
+		: _cur(cur)
+	{ }
+	
+	weight& operator*() { return *_cur; }
+
+	weights_iterator_iterator& operator++()
+	{
+		_cur++;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator_iterator operator+(T index) { return weights_iterator_iterator(_cur + index); }
+
+	template<typename T>
+	weights_iterator_iterator& operator+=(T index)
+	{
+		_cur += index;
+		return *this;
+	}
+
+	weights_iterator_iterator& operator=(const weights_iterator_iterator& other)
+	{
+		_cur = other._cur;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator_iterator& rhs) { return _cur == rhs._cur; }
+	bool operator!=(const weights_iterator_iterator& rhs) { return _cur != rhs._cur; }
+
+};
+class weights_iterator
```

Comment:
Many of my coding nitpicks apply to this weights_iterator class, too, so I won't repeat myself.   Honestly, use boost.iterator.  It will save you from having to write (and test!) all this iterator boilerplate.  It will also save others having to read an maintain it, which is always a big win.

However, let's discuss some design.  Specifically, your intention is to make it possible to iterate ranges of weights in a bucket.  As such, I would consider returning a *range* as the result of your iterator dereference - effectively, your signature would be e.g. boost::iterator_range<weight *, weight*> operator*().  This will allow you to write very clean code along the lines of:

    for (auto ww: *weight_bucket)
    {
       //
    }

There's also a downside - if you expect people to mostly access only the first element of the weight vector, you may have to write `(*iter)[0]` to access the first weight.  However, I would be cleaner to actually return a different iterator class from your `weights_vector::begin(offset)` function.  See more below.